application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/GetTrialLicenseListener.java

```diff
@@ -78,20 +78,30 @@ public void onEvent(Event event, Object source, Object data)
         trialLicenseGenerator.updateLicenses();
 
         for (ExtensionId extensionId : extensions) {
-            if (trialLicenseGenerator.canGenerateTrialLicense(extensionId)) {
-                trialLicenseGenerator.generateTrialLicense(extensionId);
-            }
+            tryGenerateTrialLicenseRecursive(extensionId);
+        }
+    }
 
-            // Since a free extension has no license to cover it's dependencies, check to see if there are any paid
-            // dependencies that need a trial license.
+    /**
+     * Try to generate a trial license for the given extension. Since a free extension has no license to cover it's
+     * dependencies, check also to see if there aren't any paid dependencies, direct or transitive, that need a trial
+     * license.
+     *
+     * @param extensionId the extension for which to generate a trial license
+     */
+    public void tryGenerateTrialLicenseRecursive(ExtensionId extensionId)
+    {
+        if (trialLicenseGenerator.canGenerateTrialLicense(extensionId)) {
+            trialLicenseGenerator.generateTrialLicense(extensionId);
+        } else {
             InstalledExtension installedExtension = installedExtensionRepository.getInstalledExtension(extensionId);
-            Collection<ExtensionDependency> dependencies = installedExtension.getDependencies();
+            if (installedExtension != null) {
+                Collection<ExtensionDependency> dependencies = installedExtension.getDependencies();
 
-            for (ExtensionDependency dependency : dependencies) {
-                ExtensionId dependencyId =
-                    new ExtensionId(dependency.getId(), dependency.getVersionConstraint().getVersion());
-                if (trialLicenseGenerator.canGenerateTrialLicense(dependencyId)) {
-                    trialLicenseGenerator.generateTrialLicense(dependencyId);
+                for (ExtensionDependency dependency : dependencies) {
+                    ExtensionId dependencyId =
+                        new ExtensionId(dependency.getId(), dependency.getVersionConstraint().getVersion());
```

Comment:
I'm not sure about this part. The version constraint specified in the POM for the dependency doesn't have to match the version that is actually installed (used at runtime). For instance, extension A might declare in its POM a dependency on extension B with version 5.3 (which means it needs **at least** version 5.3) and at runtime work with B v7.1 (because this version is needed by some other extension for instance). Moreover, you could have an extension C that specifies in its POM that B is a feature (it covers B, e.g. if B was renamed to C) and thus you can have A work with C, a different extension id than the one specified in the POM as dependency.

So I think you need to ``resolve`` the dependency id into an installed extension (i.e. to see which installed extension satisfies that dependency).